ml/notebooks/netdata_anomaly_detection_deepdive.ipynb

```diff
@@ -0,0 +1,1705 @@
+{
```

Comment:
<p>For demonstration purposes, we'll use the <a href="https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html" target="_blank">kmeans</a> from <a href="https://scikit-learn.org/stable/index.html" target="_blank">scikit-learn</a> to make the general approach as readable and understandable as possible.</p><p>In reality, the Netdata Agent uses the awesome <a href="https://github.com/davisking/dlib" target="_blank">dlib</a> C++ library and the <code><a href="http://dlib.net/ml.html#find_clusters_using_kmeans" target="_blank">find_clusters_using_kmeans</a></code> function along with a few others. You can see the Netdata KMeans code <a href="https://github.com/netdata/netdata/blob/master/ml/kmeans/KMeans.cc" target="_blank">here</a>. </p><p>The code below:</p><ol><li>Will initialize some empty objects to use during model training and inference. </li><li>Will loop over every observation and run training and inference in a similar way to how the Agent would process each observation. </li></ol><p>   Of course the Agent implementation is a lot more efficient and uses more efficient streaming and buffer-based approaches as opposed to the fairly naive implementation below.  </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/netdata/netdata/pull/12313/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='19'/>